File: 1267-remove-zero-sum-consecutive-nodes-from-linked-list/remove-zero-sum-consecutive-nodes-from-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    public ListNode removeZeroSumSublists(ListNode head) {

        ListNode dummy = new ListNode(0);
        dummy.next = head;

        HashMap<Integer, ListNode> map = new HashMap<>();

        int prefix = 0;
        ListNode curr = dummy;

        // First pass
        while (curr != null) {
            prefix += curr.val;
            map.put(prefix, curr);
            curr = curr.next;
        }

        // Second pass
        prefix = 0;
        curr = dummy;

        while (curr != null) {
            prefix += curr.val;
            curr.next = map.get(prefix).next;
            curr = curr.next;
        }

        return dummy.next;
    }
}

File: 1267-remove-zero-sum-consecutive-nodes-from-linked-list/remove-zero-sum-consecutive-nodes-from-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeZeroSumSublists(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)
        dummy.next = head

        prefix = 0
        mp = {}

        # First pass: store the last occurrence of each prefix sum
        curr = dummy
        while curr:
            prefix += curr.val
            mp[prefix] = curr
            curr = curr.next

        # Second pass: skip zero-sum sublists
        prefix = 0
        curr = dummy
        while curr:
            prefix += curr.val
            curr.next = mp[prefix].next
            curr = curr.next

        return dummy.next